vowpalwabbit/automl.cc

```diff
@@ -0,0 +1,675 @@
+      }
+      for (auto& new_exclusion : new_exclusions)
+      {
+        int config_index = static_cast<int>(configs.size());
+        exclusion_config conf(starting_budget);
+        conf.exclusions = new_exclusion;
+        configs[config_index] = conf;
+        float priority = get_priority(config_index);
+        index_queue.push(std::make_pair(priority, config_index));
+      }
+    }
+  }
+
+  // This function is triggered when all sets of interactions generated by the oracle have been tried and
+  // reached their budget. It will then add inactive configs (stored in the config manager) to the queue
+  // 'index_queue' which can be used to swap out live configs as they run out of budget. This functionality
+  // may be better within the oracle, which could generate better priorities for different configs based
+  // on ns_counter (which is updated as each example is processed)
+  bool repopulate_index_queue()
+  {
+    std::set<int> live_indices;
+    for (size_t stride = 0; stride < max_live_configs; ++stride) { live_indices.insert(scores[stride].config_index); }
+    for (auto ind_config : configs)
+    {
+      int redo_index = ind_config.first;
+      if (live_indices.find(redo_index) == live_indices.end())
+      {
+        float priority = get_priority(redo_index);
+        index_queue.push(std::make_pair(priority, redo_index));
+      }
+    }
+    return !index_queue.empty();
+  }
+
+  // This function defines the logic update the live configs' budgets, and to swap out
+  // new configs when the budget runs out.
+  void update_live_configs()
+  {
+    for (size_t stride = 0; stride < max_live_configs; ++stride)
+    {
+      int config_index = scores[stride].config_index;
+      // Special case to initialize first champion to -q ::
+      if (config_index == -2)
+      {
+        scores[stride].config_index = 0;
+        live_interactions[stride] = oc.gen_interactions(ns_counter, configs[0].exclusions);
+      }
+      if (stride == current_champ) { continue; }
+      if (config_index == -1 || configs[config_index].budget == 0)
+      {
+        if (config_index >= 0)
+        {
+          scores[stride].reset_stats();
+          configs[config_index].starting_budget *= 2;
+          configs[config_index].budget = configs[config_index].starting_budget;
+        }
+        // TODO: Add logic to erase index from configs map if wanted
+        if (index_queue.empty() && !repopulate_index_queue()) { continue; }
+        config_index = index_queue.top().second;
+        index_queue.pop();
+        scores[stride].config_index = config_index;
+        // Regenerate interactions each time an exclusion is swapped in
+        live_interactions[stride].clear();
+        live_interactions[stride] = oc.gen_interactions(ns_counter, configs[config_index].exclusions);
+        // We may also want to 0 out weights here? Currently keep all same in stride position
+      }
+      configs[config_index].budget--;
```

Comment:
We might not need to do this, if budget is compared with update_count. They might be tracking the same state right now.

Reply:
We can have a method on scored_config report back how much budget it has left by comparing budget against update_count.

Reply:
Updated. We should note that this changes the functionality from "swap out after X # of examples as non-champ" -> "swap out if not champ at example #X"